cm/cm_heat_demand/tools/geofile.py

```diff
@@ -0,0 +1,140 @@
+import json
+from os import remove, system
+from os.path import exists, isfile
+
+import numpy as np
+import rasterio
+from osgeo import gdal, osr
+
+
+def get_projection(geofile: str):
+    """Get the projection of a geofile."""
+
+    with rasterio.open(geofile) as src_file:
+        projection = src_file.crs
+    return projection
+
+
+class RasterNotOverlappedError(Exception):
+    """Exception thrown for map size error."""
+
+    pass
+
+
+def clip_raster(src: str, shapes: dict, dst: str, quiet: bool = True):
+    """
+    Take a raster and clip it according to polygon defined as a dictionary,
+    then save the result.
+
+    Inputs :
+        * src : path to the to be clipped raster.
+        * shapes : list of dictionaries representing polygons.
+        * dst : path where the clipped raster is saved.
+
+    Output :
+        * projection : projection of the clipped raster.
+    """
+
+    cutline = "cutline.json"
+    with open(cutline, "w") as file:
+        json.dump(shapes, file)
+    command = f"gdalwarp -cutline {cutline} -crop_to_cutline -dstnodata 0 {src} {dst}"
+    if quiet:
+        command += " -q"
+    system(command=command)
+    remove(cutline)
+
+    clipped_raster = read_raster(dst, return_geo_transform=False)
+    if np.max(clipped_raster) == 0:
+        raise RasterNotOverlappedError("Map return is empty.")
+
+    if not isfile(dst):
+        raise FileNotFoundError(f"The result file has not been created: {dst}")
+
+
+def read_raster(raster: str, return_geo_transform: bool = True):
+    """
+    Open the raster and return the map as a numpy array.
+
+    The geo transform is not return by default.
+    The 'return_geo_transform' must be set to True for this.
+
+    See https://gdal.org/tutorials/geotransforms_tut.html,
+    for more information about geotransform.
+
+    Inputs :
+        * raster :
+            - path to the raster.
+        * return_geo_transform (optional, set as True by default) :
+            - specifying if we want to return the geo transform.
+    """
+
+    if exists(raster):
+        geofile = gdal.Open(raster)
+        if not geofile:
+            raise Exception("Unable to load a raster")
+    else:
+        raise ValueError(f"This path does not refer to existing file : {raster}")
+
+    geo_transform = geofile.GetGeoTransform()
+    band = geofile.GetRasterBand(1)
+    array = band.ReadAsArray().astype(float)
+
+    geofile = None
+    band = None
```

Comment:
What are you doing here?

Reply:
GDAL expects us to close the dataset
https://gdal.org/api/python_gotchas.html#saving-and-closing-datasets-datasources